packages/regl-worldview/src/WorldviewContext.js

```diff
@@ -272,31 +275,51 @@ export class WorldviewContext {
     });
   }
 
-  callComponentHandlers = (objectId: number, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
-    this._hitmapCalls.forEach((_, component) => {
-      if (component instanceof Command) {
-        component.handleMouseEvent(objectId, e, ray, mouseEventName);
+  callComponentHandlers = (hitmapId: HitmapId, ray: Ray, e: MouseEvent, mouseEventName: MouseEventEnum) => {
+    this._drawCalls.forEach((drawInput, component) => {
+      if (component instanceof Command && this._hitmapIdManager.commandHasHitmapId(drawInput.instance, hitmapId)) {
+        const mouseEventObject = this.getDrawPropByHitmapId(hitmapId);
+        component.handleMouseEvent(mouseEventObject, e, ray, mouseEventName);
       }
     });
   };
 
   _drawInput = (isHitmap?: boolean) => {
-    const drawCallsMap = isHitmap ? this._hitmapCalls : this._drawCalls;
-    const sortedDrawCalls = Array.from(drawCallsMap.values()).sort((a, b) => (a.layerIndex || 0) - (b.layerIndex || 0));
+    if (isHitmap) {
+      this._hitmapIdManager.reset();
+    }
 
+    const drawCalls = Array.from(this._drawCalls.values());
+    const sortedDrawCalls = drawCalls.sort((a, b) => (a.layerIndex || 0) - (b.layerIndex || 0));
     sortedDrawCalls.forEach((drawInput: DrawInput) => {
-      const { command, drawProps, instance } = drawInput;
+      const { command, drawProps, instance, getHitmap, getActive } = drawInput;
       if (!drawProps) {
         return console.debug(`${isHitmap ? "hitmap" : ""} draw skipped, props was falsy`, drawInput);
       }
       const cmd = this._compiled.get(command);
       if (!cmd) {
-        return console.warn("could not find draw command for", instance.constructor.displayName);
+        return console.warn("could not find draw command for", instance ? instance.constructor.displayName : "Unknown");
+      }
+      // draw hitmap
+      if (isHitmap && getHitmap) {
+        const commandBoundAssignNextIds: CommandBoundAssignNextIds = (...rest) => {
+          return this._hitmapIdManager.assignNextIds(instance, ...rest);
+        };
+        const hitmapProps = getHitmap(drawProps, commandBoundAssignNextIds);
+        return cmd(hitmapProps);
+      }
+      if (getActive) {
+        const drawPropsWithActive = getActive(drawProps);
+        return cmd(drawPropsWithActive);
       }
       cmd(drawProps);
     });
   };
 
+  getDrawPropByHitmapId = (hitmapId: HitmapId): MouseEventObject => {
```

Comment:
♻️ It seems like this doesn't need to be a public function if we were to automatically call it in `readHitmap` to get the objects rather than just returning the ids?

Reply:
Refactored this to be a bit more understandable